In [1]:
import pandas as pd

# Load data
file_path = 'C:/Users/User/OneDrive - Swinburne University/Desktop/forage/bcg gen ai/financial_data.csv'
df = pd.read_csv(file_path)

# Print actual column names loaded by pandas to double-check
print("DataFrame columns after loading:")
print(df.columns)
print("-" * 30)

# Data Cleaning and Type Conversion 
# List of columns that should be numeric
numeric_cols = [
    'Total Revenue ($M)',
    'Net Income ($M)',
    'Total Assets ($M)',
    'Total Liabilities ($M)',
    'Cash Flow from Operating Activities ($M)'
]

# Apply cleaning and conversion to each numeric column
for col in numeric_cols:
    if col in df.columns: # Check if the column exists
        # Convert to string, remove commas, remove any leading/trailing spaces, then convert to numeric
        df[col] = pd.to_numeric(df[col].astype(str).str.replace(',', '').str.strip(), errors='coerce')
    else:
        print(f"Warning: Column '{col}' not found in DataFrame. Please check your CSV header names.")

# Check data types after conversion 
print("\nData Types after conversion:")
print(df[numeric_cols].dtypes)
print("-" * 30)

# Sort data before calculating percentage change 
# Ensure 'Fiscal Year' is sorted correctly for pct_change to work across years within each company
df = df.sort_values(by=['Company', 'Fiscal Year'])

# Calculate year-over-year changes 
df['Revenue Growth (%)'] = df.groupby('Company')['Total Revenue ($M)'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby('Company')['Net Income ($M)'].pct_change() * 100
df['Total Assets Growth (%)'] = df.groupby('Company')['Total Assets ($M)'].pct_change() * 100
df['Total Liabilities Growth (%)'] = df.groupby('Company')['Total Liabilities ($M)'].pct_change() * 100
df['Cash Flow from Operating Activities Growth (%)'] = df.groupby('Company')['Cash Flow from Operating Activities ($M)'].pct_change() * 100

# Fill NA values that result from pct_change calculations with 0 or an appropriate value
df.fillna(0, inplace=True)

# Display the DataFrame with new growth columns
print("\nDataFrame with Growth Metrics:")
df

DataFrame columns after loading:
Index(['Company', 'Fiscal Year', 'Total Revenue ($M)', 'Net Income ($M)',
       'Total Assets ($M)', 'Total Liabilities ($M)',
       'Cash Flow from Operating Activities ($M)'],
      dtype='object')
------------------------------

Data Types after conversion:
Total Revenue ($M)                          int64
Net Income ($M)                             int64
Total Assets ($M)                           int64
Total Liabilities ($M)                      int64
Cash Flow from Operating Activities ($M)    int64
dtype: object
------------------------------

DataFrame with Growth Metrics:


Company  Fiscal Year  Total Revenue ($M)  Net Income ($M)  \
8      Apple         2022              394328            99803   
7      Apple         2023              383285            96995   
6      Apple         2024              391035            93736   
2  Microsoft         2022              198270            72738   
1  Microsoft         2023              211915            72361   
0  Microsoft         2024              236584            78416   
5      Tesla         2022               81462            12587   
4      Tesla         2023               96773            14974   
3      Tesla         2024               97690             7153   

   Total Assets ($M)  Total Liabilities ($M)  \
8             352755                  302083   
7             352583                  290437   
6             364980                  308030   
2             364840                  198298   
1             411972                  205729   
0             484792                  253303   
5              82338                   36440   
4             106618                   43009   
3             122070                   48390   

   Cash Flow from Operating Activities ($M)  Revenue Growth (%)  \
8                                    122151            0.000000   
7                                    110543           -2.800461   
6                                    118254            2.021994   
2                                     89035            0.000000   
1                                     87582            6.882030   
0                                    109219           11.640988   
5                                     14724            0.000000   
4                                     13256           18.795267   
3                                     14923            0.947578   

   Net Income Growth (%)  Total Assets Growth (%)  \
8               0.000000                 0.000000   
7              -2.813543                -0.048759   
6              -3.359967                 3.516052   
2               0.000000                 0.000000   
1              -0.518299                12.918540   
0               8.367767                17.675959   
5               0.000000                 0.000000   
4              18.964010                29.488207   
3             -52.230533                14.492862   

   Total Liabilities Growth (%)  \
8                      0.000000   
7                     -3.855232   
6                      6.057424   
2                      0.000000   
1                      3.747390   
0                     23.124596   
5                      0.000000   
4                     18.026894   
3                     12.511335   

   Cash Flow from Operating Activities Growth (%)  
8                                        0.000000  
7                                       -9.502992  
6                                        6.975566  
2                                        0.000000  
1                                       -1.631942  
0                                       24.704848  
5                                        0.000000  
4                                       -9.970117  
3                                       12.575438

In [2]:
# Summarise these findings for each company

# Define the list of growth columns
growth_cols = ['Revenue Growth (%)', 'Net Income Growth (%)',
               'Total Assets Growth (%)', 'Total Liabilities Growth (%)',
               'Cash Flow from Operating Activities Growth (%)']

# Calculate and display Average Growth Rates per Company for all relevant metrics
print("\nAverage Growth Rates per Company (FY2022-2024):\n")
summary = df.groupby('Company')[growth_cols].mean()
summary


Average Growth Rates per Company (FY2022-2024):



Revenue Growth (%)  Net Income Growth (%)  Total Assets Growth (%)  \
Company                                                                         
Apple               -0.259489              -2.057837                 1.155764   
Microsoft            6.174339               2.616490                10.198166   
Tesla                6.580948             -11.088841                14.660357   

           Total Liabilities Growth (%)  \
Company                                   
Apple                          0.734064   
Microsoft                      8.957329   
Tesla                         10.179409   

           Cash Flow from Operating Activities Growth (%)  
Company                                                    
Apple                                           -0.842475  
Microsoft                                        7.690969  
Tesla                                            0.868440

# Financial Data Analysis for GFC AI Chatbot Project

## Data Loading and Preprocessing

Financial data was manually extracted from the latest available 10-K annual reports of Microsoft, Tesla, and Apple, directly from the SEC EDGAR database. This ensures high reliability and accuracy as these are audited financial statements. The following five core financial metrics were collected for Fiscal Years 2022, 2023, and 2024: Total Revenue, Net Income, Total Assets, Total Liabilities, and Cash Flow from Operating Activities.

The extracted data was organised into a CSV file and loaded into a pandas DataFrame. Crucially, columns containing numerical figures (such as 'Total Revenue ($M)') were cleaned by removing non-numeric characters (like commas) and then successfully converted to integer data types (`int64`), which was essential for accurate mathematical calculations and trend analysis.

---

## Key Findings by Company

### Apple
* **Performance:** A mature company showing relatively stable revenue but experiencing average slight declines in Net Income and Operating Cash Flow over the period. Revenue dipped in FY2023 but slightly recovered in FY2024. Net Income, however, saw consistent declines.
* **Growth Profile:** More modest average growth rates (some negative), reflecting its established market position. Assets and Liabilities showed minimal average growth.

### Microsoft
* **Performance:** Demonstrated strong and consistent financial performance. It achieved robust average revenue growth (6.17%) and healthy average growth across all other key metrics, including Net Income (2.62%) and Operating Cash Flow (7.69%).
* **Growth Profile:** Exhibited the most stable and balanced growth among the three companies, indicating strong operational execution and market demand.

### Tesla
* **Performance:** Characterised by rapid expansion in assets (average 14.66% growth) and liabilities (10.18%), coupled with strong average revenue growth (6.58%). However, it experienced significant volatility in Net Income, showing a large average decline (-11.09%), and variable Operating Cash Flow.
* **Growth Profile:** High-growth, but with notable swings in profitability, reflecting its dynamic and potentially more volatile industry landscape.

---

## Comparative Insights

* **Scale vs. Growth:** Apple and Microsoft operate on a significantly larger financial scale. Tesla, from a smaller base, shows higher asset and liability growth percentages indicative of aggressive expansion, but with more volatile profitability.
* **Profitability Trends:** Microsoft's profitability trends generally align with its revenue growth, while Apple saw declining net income despite stable revenue. Tesla's net income was highly volatile.
* **Cash Flow Generation:** All companies generate positive operating cash flow. Apple leads in absolute cash generated, while Microsoft shows strong, consistent growth in cash flow. Tesla's cash flow growth was more variable.

---

## Conclusion 

This analysis provides crucial insights for the GFC AI Chatbot. The chatbot can be designed to:
* **Identify company growth stages:** Distinguishing between mature (Apple), consistently growing (Microsoft), and high-growth/volatile (Tesla) profiles.
* **Analyse profitability and cash generation stability:** Flagging inconsistencies between revenue growth and net income/cash flow.
* **Facilitate comparative assessments:** Allowing users to quickly benchmark companies based on various financial metrics and growth trends, offering a nuanced view beyond just top-line numbers.

This foundational understanding of financial trends equips the chatbot with the intelligence to provide more insightful and actionable financial analysis.